### ISSUE TITLE
Conflict & serialization issues with passport and feathers-passport

### ISSUE THREAD

**Issue by startupthekid** [feathersjs/feathers-passport#8]
I posted a question on [SO earlier](http://stackoverflow.com/questions/33374672/passport-session-serialization-failed-with-serialization-functions-present) about the troubles I was having serializing my sessions using feathers passport where authentication passed and all is good but calling `req.login` failed, printing "Failed to serialize user into session". Normally, as I indicate in my post, this would be indicative of not having the serialization functions. But in my case, they are present.

I did a lot of investigating and to see if I could get session serialization working with vanilla passport, I completely removed all `require("feathers-passport)` references from my authentication files and lo and behold, auth and sessions work.

I used the setup described on the passport website:

```
app
	.configure(feathers.rest()) 
	.configure(hooks())
	.use(require('body-parser').urlencoded({ extended: true }))
	.use(require('express-session')({ secret: 'keyboard cat', resave: true, saveUninitialized: true }))
	.use(passport.initialize())
	.use(passport.session())
```

The really interesting part which took me a while to discover happened when I re-included `require("feathers-passport")`. The same authentication call that previously succeeded suddenly failed, throwing `TypeError: Cannot set property 'user' of undefined`.

There are 3 main and distinct cases that I'm able to reproduce:

1) Using `feathers-passport` and setting up the example config as is done on the website. 

In this scenario, session serialization fails (see above SO post for more details). In this scenario, the serialization fails within passport's `authenticator.js`, where `this._serializers` is `[]`. When cycling through the chain of serializers, the first is popped off and subsequently doesn't exist so the failed serialization error fails. Because the authenticator can't find any serializers, my copy of `serializeUser` never gets called (where I have a log statement)

2) Using only `passport` with no references or inclusion of `feathers-passport` module.

Here, I removed all references to feathers passport and integrated vanilla passport into my existing feathers app. Session serialization works and all is good. 

3) Same as scenario 2 except for including of `require("feathers-passport")`

Here serialization actually works. My debug statements inside the serializing and deserializing methods actually get printed except this time it fails with `TypeError: Cannot set property 'user' of undefined` in `/feathers-passport/node_modules/passport/lib/http/request.js:50:35`.

A more complete stack trace is:

```
/node_modules/feathers-passport/node_modules/passport/lib/http/request.js:50:35
/node_modules/passport/lib/authenticator.js:267:43
/node_modules/passport/lib/authenticator.js:276:7
```

What's most interesting to me is that in this scenario, it's identical to 2, using the same vanilla passport setup. The only thing that differs is the requiring of the `feathers-passport` module which leads me to believe there's some sort of library conflict going on.

I'm not quite sure what's going on here. In scenario 1, authentication works but serialization fails. Scenario 2 is all good and in 3 everything works but fails to set the user property, presumably on an undefined request.

**Comment by marshallswain** [comment]
I wonder if there is an issue of having two different versions of passport: one it your node_modules and one in the feathers-passport plugin's node_modules.  Have you tried using the `passport` option to explicitly pass in your own version?

In your server setup:
```
var passport = require('passport')
...
app.configure(feathersPassport({
    // Secret must be provided
    secret: 'feathers-rocks',
    // Also set a store
    store: store,
    resave: true,
    saveUninitialized: true,
    passport: passport
}))
```

**Comment by startupthekid** [comment]
I did try that, with these configuration options:

```
.configure(feathersPassport(function(result) {
	    var MongoStore = connectMongo(result.createSession);
	    result.secret = 'feathers-rocks';
	    result.store = new MongoStore({
	      db: 'feathers-demo'
	    });
	    result.resave = true;
	    result.saveUninitialized = true;
	    result.passport = passport;
	    return result;
	 }))
```

but unfortunately an error is getting throw from request.js inside passport, line 50:

`self._passport.session.user = obj;` where `session` is undefined. Looks like the session object isn't getting set?

**Comment by daffl** [comment]
Can you provide a breaking standalone example? I know it can be tricky sometimes to reproduce but I always find it hard looking a code snippets trying to figure out what the issue is.

**Comment by startupthekid** [comment]
```
app
	.configure(feathers.rest()) 
	.configure(hooks())
	.configure(feathersPassport(function(result) {
	    var MongoStore = connectMongo(result.createSession);
	    result.secret = 'feathers-rocks';
	    result.store = new MongoStore({
	      db: 'feathers-demo'
	    });
	    result.resave = true;
	    result.saveUninitialized = true;
	    result.passport = passport;
	    return result;
	}))
.post('/users/authenticate', function(request, response, next){
		passport.authenticate('local', function(error, user){
			request.login(user, function(error){
				if(error) return next(error);
				if(!user) return next(new Error("no user found"));
				else {
					var token = issueToken({id: user.id, type: 'user'});
					response.status(200).json({token: token});
				}
			});
		})(request, response, next);
	})
//Setup passport
var localOptions = {
	usernameField: 'email',
	passwordField: 'password'
};

passport.serializeUser(function(user, done) {
	done(null, user._id);
});

passport.deserializeUser(function(id, done) {
  User.findById(id, function (err, user) {
    done(err, user);
  });
});

passport
	.use(new LocalStrategy(localOptions, function(email, password, done){
		api.service('users').authenticateLocal(email, password, done);
	}))
```

**Comment by startupthekid** [comment]
I downgraded to passport version 0.2.2 and that fixed the issue.

So to summarize, in order to get session serialization to work, pass in the passport instance to the `feathers-passport` options and make sure that passport instance is 0.2.2.

**Comment by startupthekid** [comment]
Just a heads up btw @marshallswain: https://github.com/jaredhanson/passport/issues/400
^this issue is more talking about passport strategies depending on old versions of passport but most of the fixes were simply rolling back to version 0.2.2. Might be something similar happening with feathers here.

**Comment by daffl** [comment]
Since you can provide your own passport instance already, maybe this plugin shouldn't come with its own passport. Glad you figured it out though.
